devdoc/channel.md

```diff
@@ -0,0 +1,297 @@
+
+ - **SRS_CHANNEL_43_118: [** `channel_push` shall call `sm_exec_begin`. **]**
+
+ - **SRS_CHANNEL_43_119: [** `channel_push` shall create a `THANDLE(ASYNC_OP)` by calling `async_op_create` with `cancel_channel_op` as `cancel`. **]**
+
+ - **SRS_CHANNEL_43_120: [** `channel_push` shall store the `push_callback`, `push_context` and `data` in the `THANDLE(ASYNC_OP)`. **]**
+
+ - **SRS_CHANNEL_43_121: [** `channel_push` shall insert the created `THANDLE(ASYNC_OP)` in the list of pending operations by calling `DList_InsertTailList`. **]**
+
+ - **SRS_CHANNEL_43_122: [** set the `state` to `CHANNEL_PUSHED`. **]**
+
+ - **SRS_CHANNEL_43_123: [** set `*out_op_push` to the created `THANDLE(ASYNC_OP)`. **]**
+
+**SRS_CHANNEL_43_124: [** If the `state` of the `channel` is `CHANNEL_PULLED`: **]**
+
+ - **SRS_CHANNEL_43_125: [** `channel_push` shall call `DList_RemoveHeadList` on the list of pending operations to obtain the `operation`. **]**
+
+ - **SRS_CHANNEL_43_126: [** If the list of pending operations is now empty, `channel_push` shall set the `state` to `CHANNEL_EMPTY`. **]**
+
+ - **SRS_CHANNEL_43_127: [** `channel_push` shall set the `state` of the obtained `operation` to `CHANNEL_OP_SCHEDULED`. **]**
+
+ - **SRS_CHANNEL_43_128: [** `channel_push` shall store the `push_callback`, `push_context` and `data` in the obtained `operation`. **]**
+
+ - **SRS_CHANNEL_43_129: [** `channel_push` shall call `threadpool_schedule_work` with `execute_callbacks` as `work_function` and the obtained `operation` as `work_function_context`. **]**
+
+ - **SRS_CHANNEL_43_130: [** `channel_push` shall set `*out_op_push` to the `THANDLE(ASYNC_OP)` of the obtained `operation`. **]**
+
+**SRS_CHANNEL_43_131: [** `channel_push` shall call `srw_lock_release_exclusive`. **]**
+
+**SRS_CHANNEL_43_132: [** `channel_push` shall succeeds and return `CHANNEL_RESULT_OK`. **]**
+
+**SRS_CHANNEL_43_041: [** If there are any failures, `channel_push` shall fail and return `CHANNEL_RESULT_ERROR`. **]**
+
+
+### cancel_channel_op
+```c
+    static void cancel_channel_op(void* channel_op_context);
+```
+
+**SRS_CHANNEL_43_133: [** `cancel_channel_op` is the cancel callback that is passed to `async_op_create` when creating a `THANDLE(ASYNC_OP)` for a `channel_push` or `channel_pull` operation. **]**
+
+**SRS_CHANNEL_43_134: [** `cancel_channel_op` shall call `srw_lock_acquire_exclusive`. **]**
+
+**SRS_CHANNEL_43_135: [** If the `state` of `operation` is `CHANNEL_OP_CREATED`: **]**
+
+ - **SRS_CHANNEL_43_136: [** `cancel_channel_op` shall set the `state` to `CHANNEL_OP_CANCELLED` **]**
+
+ - **SRS_CHANNEL_43_137: [** `cancel_channel_op` shall call `DList_RemoveEntryList` to remove the `operation` from the list of pending operations. **]**
+
+ - **SRS_CHANNEL_43_138: [** `cancel_channel_op` shall call `threadpool_schedule_work` with `execute_callbacks` as `work_function` and the `operation` as `work_function_context`. **]**
```

Comment:
>`operation` as `work_function_context` [](http://example.com/codeflow?start=139&length=38)

whenever I see a `void*` pointer getting assigned some value from a THANDLE my heart skips a beat.

what happens if work has been scheduled... and work did not execute YET, but meanwhile the user decides to THANDLE_ASSIGN(ASYNC_OP) = NULL for both the pull and the push ASYNC_OPs?

Is there anybody still holding on to the ASYNC_OP? Or when the work is executed, it will find some freed memory?

Reply:
Seems to happen in other requirements too (such as SRS_CHANNEL_43_113)